kuromoji-core/src/main/java/com/atilika/kuromoji/fst/vm/Program.java

```diff
@@ -38,64 +27,56 @@ public Program() {
         this.cacheFirstIsAccept = new boolean[CACHED_CHAR_RANGE];
     }
 
-
-    public byte getOpcode(int pc) {
-        int internalIndex = pc * BYTES_PER_INSTRUCTIONS;
-        return instruction.get(internalIndex);
+    public boolean isAccept(long instruction) {
+        return (instruction & 0x0000008000000000L) != 0;
     }
 
-    public char getArg1(int pc) {
-        int internalIndex = pc * BYTES_PER_INSTRUCTIONS;
-        return instruction.getChar(internalIndex + 1);
+    public boolean isFail(long instruction) {
+        return (instruction & 0x8000000000000000L) != 0;
     }
 
-    public int getArg2(int pc) {
-        int internalIndex = pc * BYTES_PER_INSTRUCTIONS;
-        return instruction.getInt(internalIndex + 1 + 2);
+    public char getLabel(long instruction) {
+        return (char) (instruction & 0x000000000000FFFFL);
     }
 
-    public int getArg3(int pc) {
-        int internalIndex = pc * BYTES_PER_INSTRUCTIONS;
-        return instruction.getInt(internalIndex + 1 + 2 + 4);
+    public int getTargetAddress(long instruction) {
+        return isFail(instruction) ? -1 : (int) ((instruction >> 40) & 0x00000000007FFFFFL);
     }
 
-    public void addInstruction(byte op) {
-        addInstruction(op, ' ', -1, 0);
+    public int getOutput(long instruction) {
+        return (instruction & 0x0000004000000000L) != 0 ? -1 : (int) ((instruction >> 16) & 0x00000000007FFFFFL);
     }
 
-    public void addInstruction(byte op, char label, int targetAddress, int output) {
-        int currentSizePlusOneInstruction = (this.getNumInstructions() + 1) * BYTES_PER_INSTRUCTIONS;
-        if (currentSizePlusOneInstruction >= BYTES_PER_INSTRUCTIONS * numInstructionsAllocated) {
-            doubleBufferSize();
-        }
+    public void addInstruction(boolean fail, boolean accept, char label, int targetAddress, int output) {
+        assert (targetAddress < 0x7FFFFFFF);
+        assert (output < 0x7FFFFFFF);
 
-        instruction.put(op);
-        instruction.putChar(label);
-        instruction.putInt(targetAddress);
-        instruction.putInt(output);
+        // encode the instruction into a 8 byte long
+        // 3 bytes == target address for 23 bits, msb == fail
```

Comment:
One quick comment: Is it safe to assume that a target address is 0 < target_address < 2^24 = 16,777,216? I believe for the constant size dictionary (e.g. IPAdic), it should be fine. In IPAdic, the number of unique first character of a surface word is 4873. If we assume that the max length of a word in a dictionary is 3, still the worst case of FST is around 4873^3 =  115,714,886,617 (which is likely impossible to happen to be this much in practice), but I think this is worth considering if we want to use FST for a large size user dictionary.

For a long term view, one other alternative might be use delta encoding for target address, rather than storing an absolute address for a target address.